include/hitable.hpp

```diff
@@ -1,26 +1,49 @@
 #ifndef HITTABLE_H
 #define HITTABLE_H
 
-#include "ray.hpp"
+#include <variant>
+
+#include "hit_record.hpp"
+#include "material.hpp"
 #include "rtweekend.hpp"
 #include "vec.hpp"
 
-class hit_record {
+struct hittable_hit_visitor {
+ private:
+  task_context& ctx;
+  const ray& r;
+  real_t min;
+  real_t max;
+  hit_record& rec;
+  material_t& hit_material_type;
+
  public:
-  float t;         //
-  point p;         // hit point
-  vec normal;      // normal at hit point
-  bool front_face; // to check if hit point is on the outer surface
-  /*local coordinates for rectangles
-  and mercator coordintes for spheres */
-  float u;
-  float v;
-
-  // To set if the hit point is on the front face
-  void set_face_normal(const ray& r, const vec& outward_normal) {
-    front_face = dot(r.direction(), outward_normal) < 0;
-    normal = front_face ? outward_normal : vec {} - outward_normal;
+  hittable_hit_visitor(task_context& ctx, const ray& r, real_t min, real_t max,
+                       hit_record& rec, material_t& hit_material_type)
+      : ctx { ctx }
+      , r { r }
+      , min { min }
+      , max { max }
+      , rec { rec }
+      , hit_material_type { hit_material_type } {}
+
+  template <typename H> bool operator()(H&& hittable) {
+    return hittable.hit(ctx, r, min, max, rec, hit_material_type);
+  }
+
+  bool operator()(std::monostate) {
+    assert(fase && "unreachable");
+    return false;
```

Comment:
This looks complex and you always copy the elements even if they are not used by the corresponding `hit` function, compared to the previous version.
What about just
```c++
f (!dev_visit(monostate_return_false(
            [&](auto&& arg) {
              return arg.hit(ctx, r, rec1.t + 0.0001f, infinity, rec2,
                             temp_material_type);
            }),
            boundary)) {
```
and you just need to implement a single function for all the visitors
```c++
auto monostate_return_false = [](auto&& dispatch) {
  return [](auto&& arg) {
	if constexpr (std::is_same_v<std::remove_cvref_t<arg>, std::monostate>)
	  return false;
    else
      return dispatch(arg);
  }
};
```